### PR TITLE
Implementation of base route

### PR DESCRIPTION
**Pull request by kfreiman**
It still on string separator. RegExp is unsuitable: `function getBase()` returns string, wich must be same string after next using `getBase()`. 

If `loc.href.match(/^.*\/[0-9]?/)[0]` returns null, we can't construct root.
But we can do it with string separator.

If size is so important, i change 
```js
function root() {
  var origin = loc.protocol + '//' + loc.host
  var remain = loc.href.substr(origin.length)
  var i = remain.indexOf(base)
  return origin + (i > 0 ? remain.substr(0, i) : remain) + base
}
```
to
```js
function root() {
  var i = loc.href.indexOf(base)
  return (i > 0 ? loc.href.substr(0, i) : loc.href) + base
}
```
But it is not safe, because for URL like `http://ce.local/ce/df?abc=1` and developer's careless `base='/ce'`, root() returns `'http://'`

### CODE REVIEW

**Comment by cognitom** [comment]
Well..., what the `base` is..? It seems a fragment of the url. Maybe we can describe it:
> The string which match the first occurrence and divide into the base url and the path we need

or it would be better? (only allow `#` and `#!` as special cases):

> The base path or # or #!

**Comment by cognitom** [comment]
Here's RegExp case for comparison. It seems simpler for me... :worried:
> The base path (regex allowed)

Thinking about the code like this:

```javascript
var base = /^.*#/
// or var base = '/path/to/'

var href = loc.href.replace(/^.+?\/+[^\/]+/, '')
var path = href.replace(base, '')
var baseStr = href.match(base)[0] || ''
```

> If size is so important

Yeah, the size has most priority for this library. https://github.com/riot/route/issues/2#issuecomment-131691513

**Comment by kfreiman** [comment]
I separate href() function wich define string for router mainpulations. It is step to safe.
But otherwise, regexp is not useful.

**Comment by kfreiman** [comment]
this closes #4

File: lib/index.js [hunk 1]
```diff
@@ -5,18 +5,35 @@
...
+ */
+function href() {
+  return loc.href.replace(/^.+?\/+[^\/]+/, '')
+}
+
+/**
+ * Get part of current URL which contains the router params
  * @returns {string} hash string
  */
 function hash() {
```
Comment (cognitom) [inline lib/index.js]: `hash` is confusable. How about `paht()`?

File: lib/index.js [hunk 2]
```diff
@@ -54,6 +71,14 @@ route.exec = function(fn) {
 }
 
 /**
+ * Replace the default '#' base to yours,
+ * @param {string} str - string which separates app's root and router params (first match from left)
+ */
+route.base = function(str) {
+  base = str
+}
```
Comment (cognitom) [inline lib/index.js]: Allow regexp, and make `base` as string.

```javascript
/**
 * Set the base of url
 * @param {(str|RegExp)} strOrRe - a new base string or RegExp
 */
route.base = function(strOrRe) {
  base = strOrRe[0] ? strOrRe : (href().match(strOrRe) || [''])[0]
  current = path() // recalculate current path
}
```

Comment (cognitom) [inline lib/index.js]: Then we could omit `root()` func.

**Comment by cognitom** [comment]
Anyway, before merging this, we need to implement `pushState` things. Without it we can't pass the test. Can I have a time for working on it? I could be back on Sat. Thanks.

**Comment by kfreiman** [comment]
good luck!

**Comment by cognitom** [comment]
@kfreiman done.

Any feedback is helpful.
https://github.com/riot/route/blob/v2.3/lib/index.js

Finally, I decided to remove RegExp from the `route.base()`. You were right. I reallized that it could be much complicated with RegExp:

- `/path/to/somewhere#fruit/apple`
- `/path/to/somewhere#`
- `/path/to/somewhere`

Your way was much simpler. Sorry I couldn't merge your PR because I've almost rewritten from scratch. Feel free to send another PR again. Thank you.

(After adding some tests, I'll merge `v2.3` to `dev` branch, soon)

**Comment by kfreiman** [comment]
@cognitom. I'll be watching this module and Riot, and i will be happy to send PR again

**Comment by cognitom** [comment]
:v:

**Pull request closed without merging**